### PR TITLE
[Block accumulation] Remove single request execute method

### PR DESCRIPTION
**Pull request by Liorgal28**
In the current MR, I have removed the old execute method which gets all the parameters one by one, and replaced it with the one which takes a list of ExecutionRequest.

### CODE REVIEW

File: tests/simpleKVBC/TesterReplica/internalCommandsHandler.cpp [hunk 1]
```diff
@@ -74,6 +44,7 @@ void InternalCommandsHandler::execute(InternalCommandsHandler::ExecutionRequests
                 "The message is too small: requestSize is " << req.request.size() << ", required size is "
                                                             << sizeof(SimpleRequest));
       req.outExecutionStatus = -1;
+      continue;
```
Comment (teoparvanov) [inline tests/simpleKVBC/TesterReplica/internalCommandsHandler.cpp:47]: why is this necessary actually?

Comment (Liorgal28) [reply tests/simpleKVBC/TesterReplica/internalCommandsHandler.cpp:47]: If it won't be there so it will go to the next lines and get to one of the executes. we just want to go to the next request.

**Review by yuliasherman — CHANGES_REQUESTED** [review]

File: bftengine/src/preprocessor/PreProcessor.cpp [hunk 2]
```diff
@@ -701,28 +701,23 @@ uint32_t PreProcessor::launchReqPreProcessing(uint16_t clientId,
...
-                                         maxPreExecResultSize_,
-                                         (char *)getPreProcessResultBuffer(clientId),
-                                         resultLen,
-                                         replicaSpecificInfoLen,
-                                         span);
+  bftEngine::IRequestsHandler::ExecutionRequestsQueue accumulatedRequests;
+  accumulatedRequests.push_back(bftEngine::IRequestsHandler::ExecutionRequest{
+      clientId, reqSeqNum, PRE_PROCESS_FLAG, std::string(reqBuf, reqLength), std::string(maxPreExecResultSize_, 0)});
+  requestsHandler_.execute(accumulatedRequests, cid, span);
+  auto request = accumulatedRequests.back();
```
[the remark this answers is not in the document]
Comment (Liorgal28) [reply bftengine/src/preprocessor/PreProcessor.cpp]: Changed it to be explicit 'const IRequestsHandler::ExecutionRequest &'

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 3]
```diff
@@ -3644,37 +3644,35 @@ void ReplicaImp::executeReadOnlyRequest(concordUtils::SpanWrapper &parent_span,
...
-                                        request->flags(),
-                                        request->requestLength(),
-                                        request->requestBuf(),
-                                        reply.maxReplyLength(),
-                                        reply.replyBuf(),
-                                        actualReplyLength,
-                                        actualReplicaSpecificInfoLength,
-                                        span);
+    bftRequestsHandler_.execute(accumulatedRequests, request->getCid(), span);
+    single_request = accumulatedRequests.back();
```
[the remark this answers is not in the document]
Comment (Liorgal28) [reply bftengine/src/bftengine/ReplicaImp.cpp]: Done.

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 4]
```diff
@@ -3644,37 +3644,35 @@ void ReplicaImp::executeReadOnlyRequest(concordUtils::SpanWrapper &parent_span,
   uint16_t clientId = request->clientProxyId();
 
   int error = 0;
-  uint32_t actualReplyLength = 0;
```
[the remark this answers is not in the document]
Comment (Liorgal28) [reply bftengine/src/bftengine/ReplicaImp.cpp:3647]: Done.

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 5]
```diff
@@ -3644,37 +3644,35 @@ void ReplicaImp::executeReadOnlyRequest(concordUtils::SpanWrapper &parent_span,
...
-  uint32_t actualReplicaSpecificInfoLength = 0;
-
+  IRequestsHandler::ExecutionRequest single_request;
+  bftEngine::IRequestsHandler::ExecutionRequestsQueue accumulatedRequests;
+  accumulatedRequests.push_back(
+      bftEngine::IRequestsHandler::ExecutionRequest{clientId,
+                                                    static_cast<uint64_t>(lastExecutedSeqNum),
+                                                    request->flags(),
```
[the remark this answers is not in the document]
Comment (Liorgal28) [reply bftengine/src/bftengine/ReplicaImp.cpp]: Done.

File: bftengine/src/preprocessor/PreProcessor.cpp [hunk 6]
```diff
@@ -701,28 +701,23 @@ uint32_t PreProcessor::launchReqPreProcessing(uint16_t clientId,
...
-                                         (char *)getPreProcessResultBuffer(clientId),
-                                         resultLen,
-                                         replicaSpecificInfoLen,
-                                         span);
+  bftEngine::IRequestsHandler::ExecutionRequestsQueue accumulatedRequests;
+  accumulatedRequests.push_back(bftEngine::IRequestsHandler::ExecutionRequest{
+      clientId, reqSeqNum, PRE_PROCESS_FLAG, reqLength, reqBuf, std::string(maxPreExecResultSize_, 0)});
+  requestsHandler_.execute(accumulatedRequests, cid, span);
+  const IRequestsHandler::ExecutionRequest &request = accumulatedRequests.back();
+  memcpy((char *)getPreProcessResultBuffer(clientId), request.outReply.c_str(), request.outActualReplySize);
```
Comment (teoparvanov) [inline bftengine/src/preprocessor/PreProcessor.cpp:713]: This copying may turn out to be an expensive operation. Why is it necessary?

Comment (Liorgal28) [reply bftengine/src/preprocessor/PreProcessor.cpp:713]: This buffer holds the result, if we don't copy that we will lose the result and probably we use this result.
For example, I test it without copying and the preprocessor tests have failed.

Comment (teoparvanov) [reply bftengine/src/preprocessor/PreProcessor.cpp:713]: Before your changes, why didn’t we need to copy?

PS: bare in mind those buffers can hold up to 34MB of data

Comment (teoparvanov) [reply bftengine/src/preprocessor/PreProcessor.cpp:713]: Maybe we should use a raw buffer for the result too?

Comment (Liorgal28) [reply bftengine/src/preprocessor/PreProcessor.cpp:713]: we sent this buffer to the execute before my changes so we had 1 copy, since we use string to get the result so I must have another copy.

Comment (yuliasherman) [reply bftengine/src/preprocessor/PreProcessor.cpp:713]: The result should be a raw buffer, too.

Comment (Liorgal28) [reply bftengine/src/preprocessor/PreProcessor.cpp:713]: https://github.com/vmware/concord-bft/pull/1097#discussion_r545029692
That will solve the copy.

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 7]
```diff
@@ -3870,12 +3866,15 @@ void ReplicaImp::executeRequestsAndSendResponses(PrePrepareMsg *ppMsg,
...
+
     accumulatedRequests.push_back(IRequestsHandler::ExecutionRequest{
         clientId,
         static_cast<uint64_t>(lastExecutedSeqNum + 1),
         req.flags(),
-        std::string(req.requestBuf(), req.requestLength()),
-        std::string(ReplicaConfig::instance().getmaxReplyMessageSize() - sizeof(ClientReplyMsgHeader), 0),
+        req.requestLength(),
+        req.requestBuf(),
+        static_cast<uint32_t>(ReplicaConfig::instance().getmaxReplyMessageSize() - sizeof(ClientReplyMsgHeader)),
```
Comment (teoparvanov) [inline bftengine/src/bftengine/ReplicaImp.cpp:3876]: Shouldn’t this also be config_.getmaxReplyMessageSize() ?

File: bftengine/src/preprocessor/PreProcessor.cpp [hunk 8]
```diff
@@ -701,28 +701,28 @@ uint32_t PreProcessor::launchReqPreProcessing(uint16_t clientId,
...
+      bftEngine::IRequestsHandler::ExecutionRequest{clientId,
+                                                    reqSeqNum,
+                                                    PRE_PROCESS_FLAG,
+                                                    reqLength,
+                                                    reqBuf,
+                                                    maxPreExecResultSize_,
+                                                    (char *)getPreProcessResultBuffer(clientId)});
+  requestsHandler_.execute(accumulatedRequests, cid, span);
+  const IRequestsHandler::ExecutionRequest &request = accumulatedRequests.back();
+  auto status = request.outExecutionStatus;
```
Comment (teoparvanov) [inline bftengine/src/preprocessor/PreProcessor.cpp:719]: const auto status

File: bftengine/src/preprocessor/PreProcessor.cpp [hunk 9]
```diff
@@ -701,28 +701,28 @@ uint32_t PreProcessor::launchReqPreProcessing(uint16_t clientId,
...
   LOG_DEBUG(logger(), "Pre-execution operation done" << KVLOG(reqSeqNum, clientId, reqSeqNum));
-  if (status != 0 || !resultLen) {
```
Comment (teoparvanov) [inline bftengine/src/preprocessor/PreProcessor.cpp:721]: Instead of changing this code, let’s declare:

 const auto resultLen = request.outActualReplySize

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 10]
```diff
@@ -3643,45 +3643,41 @@ void ReplicaImp::executeReadOnlyRequest(concordUtils::SpanWrapper &parent_span,
...
-                                                    actualReplicaSpecificInfoLength,
-                                                    error));
-
+                                                    single_request.outActualReplySize,
+                                                    single_request.outReplicaSpecificInfoSize,
+                                                    status));
   // TODO(GG): TBD - how do we want to support empty replies? (actualReplyLength==0)
-  if (!error) {
-    if (actualReplyLength > 0) {
-      reply.setReplyLength(actualReplyLength);
```
Comment (teoparvanov) [inline bftengine/src/bftengine/ReplicaImp.cpp:3676]: In order to minimize the code changes, let’s just set values to actualReplyLength and actualReplicaSpecificInfoLength right after the call to .execute()

**Review by teoparvanov — APPROVED** [review]
LGTM, thanks Lior. 

PS: I suppose you’ve already prepared a matching MR in Concord, to make sure the bump goes smoothly?

File: bftengine/src/bftengine/ReplicaImp.cpp (hunk 10 above)
Comment (Liorgal28) [reply bftengine/src/bftengine/ReplicaImp.cpp:3676]: Done.

File: bftengine/src/bftengine/ReplicaImp.cpp (hunk 7 above)
Comment (Liorgal28) [reply bftengine/src/bftengine/ReplicaImp.cpp]: Done.

File: bftengine/src/preprocessor/PreProcessor.cpp (hunk 8 above)
Comment (Liorgal28) [reply bftengine/src/preprocessor/PreProcessor.cpp]: Done.

File: bftengine/src/preprocessor/PreProcessor.cpp (hunk 9 above)
Comment (Liorgal28) [reply bftengine/src/preprocessor/PreProcessor.cpp:721]: Done.

**Review by yuliasherman — APPROVED** [review]
LGTM
